src/main/java/org/jenkinsci/plugins/workflow/support/pickles/serialization/RiverReader.java

```diff
@@ -373,4 +375,59 @@ private static <T> T sandbox(ReadSAM<T> lambda) throws ClassNotFoundException, I
 
     }
 
+    /**
+     * Intercepts {@link StackOverflowError} and tries to record the chain of object classes which led to it.
+     */
+    private static final class TracingRiverMarshallerFactory extends RiverMarshallerFactory {
+        @Override public Unmarshaller createUnmarshaller(MarshallingConfiguration configuration) throws IOException {
+            return new RiverUnmarshaller(this, SerializableClassRegistry.getInstance(), configuration) {
+                int depth = 0;
+                StackOverflowError err;
+                List<String> trace = new ArrayList<>();
+                @Override protected Object doReadNewObject(int streamClassType, boolean unshared, boolean discardMissing) throws ClassNotFoundException, IOException {
+                    depth++;
+                    Object o;
+                    try {
+                        o = super.doReadNewObject(streamClassType, unshared, discardMissing);
+                    } catch (StackOverflowError x) {
+                        // Will cause a StreamCorruptionError eventually, but not before we capture the parent objects:
+                        o = null;
+                        err = x;
```

Comment:
From some local testing, I think swallowing the `StackOverflowError` like this also has the side effect of making the standard tracing mechanism in JBoss marshalling work, at least to some extent, because it seems to have around 100 fewer entries than the custom trace added in this PR. For example, along with the new trace I also see the following, which does not exist without the PR:

```
Caused by: an exception which occurred:
	in field com.cloudbees.groovy.cps.impl.BlockScopeEnv.types
	in object com.cloudbees.groovy.cps.impl.BlockScopeEnv@2ffb8577
	in object of type com.cloudbees.groovy.cps.impl.BlockScopeEnv
	in field com.cloudbees.groovy.cps.impl.ProxyEnv.parent
	in object com.cloudbees.groovy.cps.impl.TryBlockEnv@64d0aa4f
	in object of type com.cloudbees.groovy.cps.impl.TryBlockEnv
	in field com.cloudbees.groovy.cps.impl.ProxyEnv.parent
	in object com.cloudbees.groovy.cps.impl.BlockScopeEnv@17735b9f
	in object of type com.cloudbees.groovy.cps.impl.BlockScopeEnv
	in field com.cloudbees.groovy.cps.impl.ProxyEnv.parent
        ...
```

At one point I tried to augment the CPS `Env` objects to carry a `SourceLocation` that we could then use to augment JBoss Marshalling trace information to make it easier to understand where the objects are coming from, but I never managed to get things to work.

Reply:
I think that trace is due to insertion of `null` into a place where an `Object` is expected, so a secondary error that we intentionally generate as the price of being able to continue walking back up the stack and collecting returning values from `readObject`. AFAICT it tells you nothing about the primary `StackOverflowError`.

The basic problem is that given the `protected` methods available in `RiverUnmarshaller` (`readFields` does not count because one of its args is package-scope!) all we can do is record what `readObject` successfully returns, but we cannot intercept what class it is _starting_ to load. So the only thing we can do is attempt to force hundreds of `readObject` stack frames to return normally, noting the class name being loaded in each, before going back and marking the whole deserialization as a failure. This is clearly inferior to what you would be able to get from, say, a parser event listener with some kind of `startElement` method.